### PR TITLE
Feat: better Marshaling

### PR DESCRIPTION
**Pull request by hsanjuan**
This PR revisits the marshaling part of go-libp2p-raft.

Back in the day I had horrible problems getting things to deserialize correctly onto structures.

The encoding methods used byte slices as input/output which meant that snapshotting had to
make a full copy of the state into memory before being able to write the snapshot (to the network,
or to disk), and also before restoring it. This is quite innefficient for big states.

Changing the Marshaling approach to take readers/writers directly fixes that, and it turns out that it also
fixes many other issues that had me include things like `stateWrapper` and `opWrapper` somehow or another things did not work correctly.

The changes here are breaking in some ways:

* Non-pointer operations and states are really not working anymore (before it was suggested that they shoold be pointers but things kind of worked because they were wrapped with pointer objects).
* Default snapshot format changes slightly (no more wrapper in `stateWrapper`), therefore probably old snapshots cannot be deserialized anymore.
  * This does not affect cluster because we use our own Marshaling.
* `MarshableState` interface is now `Marshable`. Method signature has changed and it is now also used for operations. This allows users to use the most efficient serializing for their operations.

### CODE REVIEW

**Comment by hsanjuan** [comment]
@lanzafame: can you review this? i don't think you are in the libp2p org or on the libp2p/go team, can you ask to be added?

@kishansagathiya you too should be added to libp2p. Might be interesting that you get familiar with this library too.

cc @mgoelzer can probably help.

**Comment by hsanjuan** [comment]
Maybe we should handle https://golang.org/pkg/encoding/#BinaryMarshaler rather than creating our own `Marshable` interface. But that would mean no context propagation and no tracing (which I think is fine).

File: fsm.go [hunk 1]
```diff
@@ -108,27 +95,19 @@ func (fsm *FSM) Snapshot() (raft.FSMSnapshot, error) {
...
 
-	// Encode the state
-	bytes, err := EncodeSnapshot(fsm.stateWrap.State)
-	if err != nil {
-		return nil, err
-	}
-
-	var snap fsmSnapshot = bytes
-	return snap, nil
+	return &fsmSnapshot{
```
Comment (lanzafame) [inline fsm.go]: Could we one line this struct so the second return value is more obvious please?

File: fsm.go [hunk 2]
```diff
@@ -184,16 +163,21 @@ func (fsm *FSM) updateSubscribers() {
...
+func (snap *fsmSnapshot) Persist(sink raft.SnapshotSink) error {
+	snap.fsm.mux.Lock()
+	defer snap.fsm.mux.Unlock()
+	err := EncodeSnapshot(snap.fsm.state, sink)
 	if err != nil {
+		sink.Cancel()
 		return err
 	}
-	return sink.Close()
+	sink.Close()
```
Comment (lanzafame) [inline fsm.go]: Why not return any potential error from `sink.Close()`?

File: consensus.go [hunk 3]
```diff
@@ -40,19 +45,37 @@ type stateOp struct {
...
 //
 // Note that this initial state is not agreed upon in the cluster and that
 // GetCurrentState() will return an error until a state is agreed upon. Only
 // states submitted via CommitState() are agreed upon.
+//
+// The state can optionally implement the Marshable interface. The
+// methods will be used to serialize and deserialize Raft snapshots. By
+// default, the state is [de]serialized using Msgpack.
+//
+// We recommend using OpLog when handling very big states, as otherwise the
```
[the remark this answers is not in the document]
Comment (hsanjuan) [reply consensus.go]: Yes, you can use `CommitState` (so, committing states rather than Ops).

File: fsm.go (hunk 2 above)
Comment (hsanjuan) [reply fsm.go]: hmm I have no idea

**Comment by hsanjuan** [comment]
> Just some minor cosmetic fixes and clarifications.

Thanks, I think I have addressed all comments.

**Pull request merged**